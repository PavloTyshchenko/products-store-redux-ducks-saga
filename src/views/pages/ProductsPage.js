import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { Row, Col } from 'react-bootstrap';

import Sidebar from '../components/sidebar/Sidebar';
import Products from '../components/products/Products';

import * as productDuck from '../../ducks/products';

const ProductsPage = (props) => {

    const {
        getProducts,
        searchProducts,
        showClear,
        clearProducts
    } = props;

    const { category, term } = props.match.params;

    useEffect(() => {

        // check with params category and term - if they are - dispatch action

        if (category && !term) {
            searchProducts(category, undefined);
        };

        if (category && term) {
            searchProducts(category, term);
        };

        getProducts();

        // eslint-disable-next-line
    }, []);

 
    const clearProductsHandler = () => {
        props.history.push(`/products/All`);
        clearProducts()
    }

    return (
        <Row>
            <Col md={3}>
                <Sidebar showClear={showClear} clearProducts={clearProductsHandler} />
            </Col>

            <Col md={9} >
                <Products />
            </Col>
        </Row>
    );
};

function mapStateToProps(state) {
    return {
        showClear: state.products.showClear
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        getProducts: () => {
            dispatch(productDuck.productsRequested());
        },
        clearProducts: () => {
            dispatch(productDuck.clearProducts());
        },
        searchProducts: (category, search_term) => {
            dispatch(productDuck.searchProducts(category, search_term));
        }
    };
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(ProductsPage));
