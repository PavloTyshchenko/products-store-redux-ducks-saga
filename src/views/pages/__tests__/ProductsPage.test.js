import React from 'react';
import ReactDOM from 'react-dom';
import { act } from 'react-dom/test-utils';
import { shallow } from 'enzyme';

import { ProductsPage } from '../ProductsPage';

const defaultProps = {
    getProducts: jest.fn(),
    searchProducts: jest.fn(),
    showClear: false,
    clearProducts: jest.fn(),
    match: {
        params: {
            category: 'All',
            term: ''
        }
    }
};

const setup = (propsOption = {}) => {
    const props = Object.assign(defaultProps, propsOption);
    const wrapper = shallow(<ProductsPage {...props} />);
    return {
        wrapper,
        props
    };
};

describe('TEST Products page', () => {

    test('Should render without errors', () => {
        const { wrapper } = setup();
        const component = wrapper.find(`[data-test='productsPage']`);
        expect(component.length).toBe(1);
    });

    test('Should match snapshot', () => {
        const { wrapper } = setup();
        expect(wrapper.debug()).toMatchSnapshot();
    });

    // describe('TEST Logic', () => {

    //     describe('TEST useEffect', () => {

    //         test('Should call getProducts first', () => {
    //             let container;
    //             container = document.createElement('div');
    //             document.body.appendChild(container);
    //             act(() => {
    //                 ReactDOM.render(<ProductsPage {...defaultProps} />, container);
    //             })
    //             const { wrapper, props } = setup();
    //             expect(props.getProducts).toHaveBeenCalledTimes(1);
    //         });
    //     });
    // });
});