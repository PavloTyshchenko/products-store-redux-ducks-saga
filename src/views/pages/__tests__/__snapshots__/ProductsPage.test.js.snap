// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TEST Products page Should match snapshot 1`] = `
"<Bootstrap(Row) data-test=\\"productsPage\\">
  <Col md={3}>
    <Sidebar showClear={false} clearProducts={[Function: clearProductsHandler]} />
  </Col>
  <Col md={9}>
    <Connect(Products) />
  </Col>
</Bootstrap(Row)>"
`;
